### DISCUSSION TITLE
Permissions

### DISCUSSION THREAD

**Discussion by kotyara85** [rabbitmq/rabbitmq-server#3890] [Questions] [answered]
Hello there,
I created a new `test` used and denied all access for it
![Screen Shot 2021-12-13 at 10 06 33 AM](https://user-images.githubusercontent.com/10145395/145865084-ced539bc-9bda-460e-9ccb-53318b19006c.png)

It still can publish messages to `/` host and direct exchange.
Is there something I'm missing?

code
```
conn, err := amqp.Dial("amqps://test:password@localhost:5671/")
failOnError(err, "Failed to connect to RabbitMQ")
defer conn.Close()

ch, err := conn.Channel()
failOnError(err, "Failed to open a channel")
defer ch.Close()

err = ch.Publish(
	"direct", // exchange
	"key",     // routing key
	false,        // mandatory
	false,        // immediate
	amqp.Publishing{
		ContentType: "text/plain",
		Body:        []byte("test")
		Headers:     nil,
	})

failOnError(err, "Failed to publish a message")
log.Printf(" [x] Sent %s", time.Since(s))
```

**Comment by lukebakken** [comment]
Hello! Please export the definitions from this test node and _attach_ them to your response.

**Reply by kotyara85** [reply]
Here it's

[definitions.json.txt](https://github.com/rabbitmq/rabbitmq-server/files/7808245/definitions.json.txt)

Thank you

**Comment by lukebakken** [comment]
@kotyara85 - In the future, provide a complete code repository that can be cloned and run. Right now I'm wasting time getting my environment set up to run your code. I have to remember how to set up a Node package, etc. This may seem simple to you but remember that the people who maintain RabbitMQ (and other software) have to switch between development environments to assist users.

In fact it's not immediately obvious which Node package you're using, so I will have to test with Python. It doesn't make a difference in this case.

**Comment by lukebakken** [comment, marked as answer]
I get the following when I try to publish using Pika:

```
WARNING:pika.channel:Received remote Channel.Close (403): "ACCESS_REFUSED - access to exchange 'direct' in vhost '/' refused for user 'test'" on <Channel number=1 CLOSING conn=<SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncPlaintextTransport object at 0x0000018B7A1A1A80> params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>
```

In the RabbitMQ logs:

```
2022-01-04 07:28:22.032000-08:00 [error] <0.3373.0> operation basic.publish caused a channel exception access_refused: access to exchange 'direct' in vhost '/' refused for user 'test'
```

I suggest checking your RabbitMQ log files. The javascript library you're using may not be propagating the permission error correctly.

If you'd like to provide a complete set of code I can clone and run, I can confirm.

**Reply by kotyara85** [reply]
I’m not using node, it’s golang 1.17
Anyway, thanks for looking into

**Reply by lukebakken** [reply]
> I’m not using node, it’s golang 1.17

Ha! That explains it. It looked similar enough at first glance. If you provide a repo I can clone it's easy for me to run that code.
